### PR TITLE
Feature/157 Full-text-search for local knowledge base with LuceneSail

### PR DESCRIPTION
**Pull request by rcffc**
- [x] db changeset that adds the `fullTextSearchIri` needs to provide a default value for existing KBs, otherwise it cannot be applied. It should be changed such that the column is first created nullable, then we run SQL commands to set the the value to the Virtuoso IRI if `supportConceptLinking` is true and to the value that marks that FTS is not supported if `conceptLinking` is false. After this, the column can be set to non-nullable. Thinking about it, we might also just leave the value nullable and take a null value as FTS not being supported.

### CODE REVIEW

**Comment by reckart** [comment]
Lots of conflicts. This branch needs attention before it gets completely stale (if it is not already).

**Comment by reckart** [comment]
For #19, I tried upgrading to 2.4.0-M3 - but with that version, I could not add new KBs in INCEpTION anymore. Seems to me to be a bug in the RDF4J milestone - need to double-check. I expect this will bite us as well when we try using our lucene storage fork which is also based on the 2.4.x version.

File: inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/KnowledgeBaseServiceImpl.java [hunk 1]
```diff
@@ -1140,47 +1194,95 @@ else if (!stmts.hasNext()) {
...
+            log.info("LuceneIndex updated by creating Concept with id [{}] and label [{}].", aId,
+                aLabel);
+        }
+        catch (IOException e) {
+            log.error("Could not add concept with id [{}] and label [{}] to LuceneIndex.", aId,
+                aLabel);
+        }
+    }
+
+    private void indexUpdatedConcept(String aId, String aLabel, IndexWriter aIndexWriter)
```
Comment (reckart) [inline inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/KnowledgeBaseServiceImpl.java]: We shouldn't have to directly interact with the index at all. It should be entirely sufficient to wrap the native `SailRepositoryConfig` with a `LuceneSailConfig` and then using the additional properties which this sail makes accessible in SPARQL queries: http://docs.rdf4j.org/programming/#_full_text_indexing_with_the_lucene_sail

File: inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/yaml/KnowledgeBaseMapping.java [hunk 2]
```diff
@@ -54,8 +54,10 @@
 
     @JsonProperty("property-description")
     private IRI propertyDescriptionIri;
-    
-    
+
+    @JsonProperty("full-text-search")
+    private IRI ftsIri;
```
[the remark this answers is not in the document]
Comment (rcffc) [reply inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/yaml/KnowledgeBaseMapping.java]: I had planned to inject a different full-text matching part into the query, depending on kb.getFtsIri()

File: inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/KnowledgeBaseServiceImpl.java [hunk 3]
```diff
@@ -485,6 +546,14 @@ public void updateProperty(KnowledgeBase kb, KBProperty aProperty)
     public void deleteProperty(KnowledgeBase kb, KBProperty aType)
     {
         getReificationStrategy(kb).deleteProperty(kb, aType);
+        if (kb.getType().equals(RepositoryType.LOCAL)) {
+            try {
+                indexDeletedConcept(aType.getIdentifier(), aType.getName(), getIndexWriter(kb));
```
[the remark this answers is not in the document]
Comment (rcffc) [reply inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/KnowledgeBaseServiceImpl.java]: What do you mean with transparently?
I have tested it without manual indexing, and the label of the KBObject is still selectable after deleting it from the KB.

**Comment by rcffc** [comment]
I have a problem with getting full-text search to work, maybe you can have a look @reckart 
 @Rentier 
https://stackoverflow.com/questions/52368361/lucene-full-text-search-only-works-for-labels-that-match-search-string-exactly

File: inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/yaml/KnowledgeBaseMapping.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (reckart) [reply inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/yaml/KnowledgeBaseMapping.java]: What would you do if the user enters an IRI that you don't support? Or would the UI only allow to choose between `bif:contains` and `search:matches`?

File: inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/KnowledgeBaseServiceImpl.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (reckart) [reply inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/KnowledgeBaseServiceImpl.java]: It seems to me that you are creating a second Lucene index in addition to the one that the LuceneSail creates and manages. Why?

[the remark this answers is not in the document]
Comment (rcffc) [reply inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/KnowledgeBaseServiceImpl.java]: From what I have experienced the Sail is not aware of new KBObjects added/updated/deleted, so I have to notify it manually.

File: inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/yaml/KnowledgeBaseMapping.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (rcffc) [reply inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/yaml/KnowledgeBaseMapping.java]: It seems like other RDF stores like Jena and YARS also support the Virtuoso-like query. (http://www.l3s.de/~minack/rdf-fulltext-benchmark/) So the Virtuoso part could be the default, and if a special IRI like Lucene is used, a different query part can be injected.

**Comment by rcffc** [comment]
@reckart I have added a test.

File: inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/service/ConceptLinkingService.java [hunk 4]
```diff
@@ -100,15 +106,28 @@
...
+    {
+
+    }
+
+    public ConceptLinkingService(KnowledgeBaseService aKbService,
+        EntityLinkingProperties aProperties)
+    {
+        kbService = aKbService;
+        properties = aProperties;
+        stopwordsFile = new File("${repository.path}/resources/stopwords-en.txt");
```
Comment (reckart) [inline inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/service/ConceptLinkingService.java]: The repository path is not substituted here. `${...}` does not work when used as a parameter to "File"

File: inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/util/QueryUtil.java [hunk 5]
```diff
@@ -215,58 +213,52 @@ public static TupleQuery generateCandidateFullTextQuery(RepositoryConnection con
...
-                + (aKb.getDefaultLanguage() != null ? aKb.getDefaultLanguage() : "en") + "\" )",
+            "  ?e2 ?labelIri ?label.",
             "}");
 
         ValueFactory vf = SimpleValueFactory.getInstance();
 
         TupleQuery tupleQuery = conn.prepareTupleQuery(QueryLanguage.SPARQL, query);
-        tupleQuery.setBinding("string", vf.createLiteral(aString));
+        // add wildcard '*' to perform wildcard search
+        tupleQuery.setBinding("string", vf.createLiteral(aString + "*"));
```
Comment (reckart) [inline inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/util/QueryUtil.java]: I guess for Virtuoso, the `*` should not be added, right?

File: inception-kb/pom.xml [hunk 6]
```diff
@@ -58,8 +63,17 @@
       <artifactId>commons-compress</artifactId>
     </dependency>
     <dependency>
-      <groupId>org.eclipse.rdf4j</groupId>
+      <groupId>org.apache.lucene</groupId>
+      <artifactId>lucene-core</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.lucene</groupId>
+      <artifactId>lucene-spatial-extras</artifactId>
```
Comment (reckart) [inline inception-kb/pom.xml]: Why do we need the `spatial` dependency?

File: inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/KnowledgeBaseServiceImpl.java [hunk 7]
```diff
@@ -114,6 +124,9 @@
     @org.springframework.beans.factory.annotation.Value(value = "${data.path}/kb")
     private File dataDir;
 
+    @org.springframework.beans.factory.annotation.Value(value = "${repository.path}/luceneIndex")
```
Comment (reckart) [inline inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/KnowledgeBaseServiceImpl.java]: The KB lucene indexes should be stored next to the KB stores, i.e. under `${repository.path}/kb`.

File: inception-kb/src/main/resources/de/tudarmstadt/ukp/inception/kb/model/db-changelog.xml [hunk 8]
```diff
@@ -331,4 +331,35 @@
...
+              defaultValue="http://www.w3.org/2000/01/rdf-schema#comment">
+        <constraints nullable="false" />
+      </column>
+    </addColumn>
+  </changeSet>
+
+  <changeSet author="INCEpTION Team" id="20180511">
+    <preConditions onFail="MARK_RAN">
+      <not>
+        <columnExists tableName="knowledgebase" columnName="ftsIri"/>
```
Comment (reckart) [inline inception-kb/src/main/resources/de/tudarmstadt/ukp/inception/kb/model/db-changelog.xml]: Mind that the project importer needs to be changed as well when a new non-null IRI is added because previously exported projects do not contain the IRI and a default IRI needs to be set on import.

File: inception-kb/src/main/resources/de/tudarmstadt/ukp/inception/kb/model/db-changelog.xml [hunk 9]
```diff
@@ -331,4 +331,35 @@
...
+    <addColumn tableName="knowledgebase">
+      <column name="descriptionIri"
+              type="VARCHAR(255)"
+              defaultValue="http://www.w3.org/2000/01/rdf-schema#comment">
+        <constraints nullable="false" />
+      </column>
+    </addColumn>
+  </changeSet>
+
+  <changeSet author="INCEpTION Team" id="20180511">
```
Comment (reckart) [inline inception-kb/src/main/resources/de/tudarmstadt/ukp/inception/kb/model/db-changelog.xml]: `20180511` -> 11th May 2018. That's long past!

File: inception-kb/src/main/resources/de/tudarmstadt/ukp/inception/kb/model/db-changelog.xml [hunk 10]
```diff
@@ -331,4 +331,35 @@
...
+    <preConditions onFail="MARK_RAN">
+      <not>
+        <columnExists tableName="knowledgebase" columnName="ftsIri"/>
+      </not>
+    </preConditions>
+
+    <addColumn tableName="knowledgebase">
+      <column name="ftsIri"
+              type="VARCHAR(255)">
```
Comment (reckart) [inline inception-kb/src/main/resources/de/tudarmstadt/ukp/inception/kb/model/db-changelog.xml]: That should be non-null, right? Or can it be null in which case full-text search (i.e. concept linking) would not be supported and we could remove the "supports concept linking" checkbox?

File: inception-kb/src/main/resources/de/tudarmstadt/ukp/inception/kb/model/db-changelog.xml [hunk 11]
```diff
@@ -331,4 +331,35 @@
       </column>
     </addColumn>
   </changeSet>
+
+  <changeSet author="INCEpTION Team" id="20180503">
```
Comment (reckart) [inline inception-kb/src/main/resources/de/tudarmstadt/ukp/inception/kb/model/db-changelog.xml]: Why did you add a duplicate of changeset `20180503` to the changelog?

File: inception-ui-kb/src/main/java/de/tudarmstadt/ukp/inception/ui/kb/project/KnowledgeBaseDetailsPanel.html [hunk 12]
```diff
@@ -208,26 +215,27 @@ <h3 wicket:id="name" class="panel-title"></h3>
...
-              <input wicket:id="enabled" type="checkbox" /> <wicket:label key="enabled" />
-            </label>
-          </div>
-          <div class="checkbox">
-            <label wicket:for="supportConceptLinking">
-              <input wicket:id="supportConceptLinking" type="checkbox" />
-                <wicket:label key="kb.supportConceptLinking" />
-            </label>
-          </div>
+        <label class="col-sm-2 control-label">
```
Comment (reckart) [inline inception-ui-kb/src/main/java/de/tudarmstadt/ukp/inception/ui/kb/project/KnowledgeBaseDetailsPanel.html]: What is the purpose of the changes in this block?

File: inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/KnowledgeBaseServiceImpl.java [hunk 13]
```diff
@@ -1184,4 +1197,57 @@ public File readKbFileFromClassPathResource(String aLocation) throws IOException
...
+    private IndexWriter getIndexWriter(KnowledgeBase aKb) throws IOException
+    {
+        Analyzer analyzer = new StandardAnalyzer();
+        Directory directory = FSDirectory
+            .open(new File(luceneIndexDir, aKb.getRepositoryId()).toPath());
+        return new IndexWriter(directory, new IndexWriterConfig(analyzer));
+    }
+
+    @Override
+    public void indexLocalKb(KnowledgeBase aKb)
```
Comment (reckart) [inline inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/KnowledgeBaseServiceImpl.java]: There is already `LuceneSail.reindex()`. I expect it should be possible to do something like `((LuceneSail) ((SailRepository) repoManager.getRepository(kb.getRepositoryId())).getSail()).reindex()` - obviously not like that but rather with a could of "if / instanceof" checks along the way.

File: inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/KnowledgeBaseServiceImpl.java [hunk 14]
```diff
@@ -1184,4 +1197,57 @@ public File readKbFileFromClassPathResource(String aLocation) throws IOException
...
+                }
+                conn.commit();
+            }
+        }
+        catch (IOException e) {
+            log.error("Could not index local KB.", e);
+        }
+    }
+
+    private void indexCreatedConcept(String aId, String aLabel, IndexWriter aIndexWriter)
```
Comment (reckart) [inline inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/KnowledgeBaseServiceImpl.java]: Did you debug if `LuceneSailConnection.commit()` properly updates the index with the stored triples?

File: inception-kb/src/main/resources/de/tudarmstadt/ukp/inception/kb/model/db-changelog.xml (hunk 8 above)
Comment (rcffc) [reply inception-kb/src/main/resources/de/tudarmstadt/ukp/inception/kb/model/db-changelog.xml]: Where do I find the class that needs to be changed?

File: inception-ui-kb/src/main/java/de/tudarmstadt/ukp/inception/ui/kb/project/KnowledgeBaseDetailsPanel.html (hunk 12 above)
Comment (rcffc) [reply inception-ui-kb/src/main/java/de/tudarmstadt/ukp/inception/ui/kb/project/KnowledgeBaseDetailsPanel.html]: Moving the "supportsConceptLinking"  checkbox to the "Type" step.

**Comment by rcffc** [comment]
I have addressed the points that you mentioned.

File: inception-kb/src/main/resources/de/tudarmstadt/ukp/inception/kb/model/db-changelog.xml (hunk 8 above)
Comment (reckart) [reply inception-kb/src/main/resources/de/tudarmstadt/ukp/inception/kb/model/db-changelog.xml]: `de.tudarmstadt.ukp.inception.kb.exporter.KnowledgeBaseExporter` and `de.tudarmstadt.ukp.inception.kb.exporter.KnowledgeBaseExporterTest`

**Comment by reckart** [comment]
@rcffc please fix the JavaDoc (pointy brackets need to be written as HTML entities in JavaDoc):

```
[ERROR] /home/svc_jenkins/workspace/INCEpTION (GitHub) (PR builder)/inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/model/KnowledgeBase.java:117: error: malformed HTML
[ERROR] * The IRI used for full text search, e.g. bif:contains or <http://www.openrdf.org/contrib/lucenesail#>
[ERROR] ^
[ERROR] /home/svc_jenkins/workspace/INCEpTION (GitHub) (PR builder)/inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/model/KnowledgeBase.java:117: error: bad use of '>'
[ERROR] * The IRI used for full text search, e.g. bif:contains or <http://www.openrdf.org/contrib/lucenesail#>
[ERROR] ^
```

Please also see if checkstyle is ok with the code.

**Comment by rcffc** [comment]
@reckart Can you please have another look?

File: inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/service/ConceptLinkingService.java [hunk 15]
```diff
@@ -455,15 +496,15 @@ private String tokensToString(List<Token> aSentence)
         for (Token t : aSentence) {
             joiner.add(t.getCoveredText());
         }
-        return joiner.toString().substring(0, joiner.length() - 1);
+        return joiner.toString().substring(0, (joiner.length() != 0) ? joiner.length() - 1 : 0);
```
Comment (Rentier) [inline inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/service/ConceptLinkingService.java]: Can you add a comment why you need to do this?

File: inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/KnowledgeBaseServiceImpl.java [hunk 16]
```diff
@@ -254,8 +270,10 @@ public RepositoryImplConfig getNativeConfig()
         // See #221 - Disabled because it is too slow during import
         // return new SailRepositoryConfig(
         //   new ForwardChainingRDFSInferencerConfig(new NativeStoreConfig()));
-        
-        return new SailRepositoryConfig(new NativeStoreConfig());
+
+        LuceneSailConfig config = new LuceneSailConfig(new NativeStoreConfig());
+        config.setIndexDir("indexDir");
```
Comment (Rentier) [inline inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/KnowledgeBaseServiceImpl.java]: Why dont you set it to `luceneIndexDir`?

File: inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/KnowledgeBaseServiceImpl.java (hunk 14 above)
Comment (Rentier) [reply inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/KnowledgeBaseServiceImpl.java]: Where is this function used?

File: inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/exporter/ExportedKnowledgeBase.java [hunk 17]
```diff
@@ -195,6 +198,16 @@ public void setPropertyDescriptionIri(String aPropertyDescriptionIri)
...
     }
 
+    public String getFullTextSearchIri()
+    {
+        return fullTextSearchIri;
+    }
+
+    public void setFullTextSearchIri(String fullTextSearchIri)
+    {
+        this.fullTextSearchIri = fullTextSearchIri;
```
Comment (Rentier) [inline inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/exporter/ExportedKnowledgeBase.java]: aFullTextSearchIri

File: inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/service/ConceptLinkingService.java [hunk 18]
```diff
@@ -255,14 +281,17 @@ public void init()
...
 
                 CandidateEntity newEntity = new CandidateEntity(
                     (e2 != null) ? e2.stringValue() : "",
                     (label != null) ? label.stringValue() : "",
-                    (altLabel != null) ? altLabel.stringValue()
+                    (altLabel != null) ? altLabel.stringValue() :
                         // Exact matching does not use altLabel
-                        : (label != null) ? label.stringValue() : "",
-                    (description != null) ? description.stringValue() : "");
+                        (label != null) ? label.stringValue() : "",
```
[the remark this answers is not in the document]
Comment (rcffc) [reply inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/service/ConceptLinkingService.java]: It is indented.

File: inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/service/ConceptLinkingService.java [hunk 19]
```diff
@@ -344,31 +373,40 @@ private synchronized Sentence getMentionSentence(JCas aJcas, int aBegin)
...
         Set<String> sentenceContentTokens = new HashSet<>();
-        for (Token t : JCasUtil.selectCovered(Token.class, mentionSentence)) {
-            boolean isNotPartOfMention = !splitMention.contains(t.getCoveredText());
-            boolean isNotStopword = (stopwords == null) || (stopwords != null && !stopwords
-                .contains(t.getCoveredText().toLowerCase(Locale.ENGLISH)));
-            if (isNotPartOfMention && isNotStopword) {
-                sentenceContentTokens.add(t.getCoveredText().toLowerCase(Locale.ENGLISH));
+        List<Token> mentionContext = new ArrayList<>();
+
+        if (aJCas != null) {
```
[the remark this answers is not in the document]
Comment (rcffc) [reply inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/service/ConceptLinkingService.java]: For testing, I just pass a "null" for the CAS argument.

**Comment by rcffc** [comment]
For some reason, the Dashboard Page doesn't show the KB page after creation. There is no error, the last messages after creation of a local KB are;

```
2018-10-02 14:04:44 DEBUG [admin] FileUploadDownloadHelper - Creating temporary file for [more_pets.ttl] in [/tmp/inception_file8971659223205485340more_pets.ttl]
2018-10-02 14:04:49 DEBUG [admin] KnowledgeBaseServiceImpl - Stream is not compressed, continue as is.

```
I do not encounter this problem on master @ aea8c297d32cf1e4bdbde03a158749ca91ef2835.

**Comment by reckart** [comment]
@rcffc if you have a directory called "SYSTEM" in your `INCEPTION_HOME/kb/repositories` folder, try renaming it to `SYSTEM.off`. Might be the RDF4J bug.

**Comment by reckart** [comment]
@rcffc actually, it's a bug in the menu item - I'm looking into it

**Comment by reckart** [comment]
@rcffc I believe the menu item bug should be fixed. If not, please let me know. Could you please update the branch, resolve the conflicts, and check that it builds properly on Jenkins?

**Comment by rcffc** [comment]
When I create a new KB, I now get:
```
2018-10-09 15:06:35 ERROR [admin] DefaultExceptionMapper - Unexpected error occurred
org.apache.wicket.WicketRuntimeException: Error attaching this container for rendering: [ListItem [Component id = 0]]
	at org.apache.wicket.MarkupContainer.onBeforeRenderChildren(MarkupContainer.java:1765) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.Component.onBeforeRender(Component.java:3788) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.Component.beforeRender(Component.java:937) ~[wicket-core-8.1.0.jar:8.1.0]
...
Caused by: java.lang.IllegalStateException: onConfigure() in the hierarchy of de.tudarmstadt.ukp.inception.ui.kb.project.wizard.KnowledgeBaseCreationWizard$LocalRepositoryStep$1$1 has not called super.onConfigure() at org.apache.wicket.Component.verifySuperCall(Component.java:1013) ~[wicket-core-8.1.0.jar:8.1.0]
...
```
Furthermore, the Title Square looks a bit strange:
selection_098
Experienced on Inception @ d976ec7678ab9d5a036f77defd2faeb425a1702e, Webanno @ 1d54510b25a72c0086e5c598c1898357b317431

**Comment by reckart** [comment]
@rcffc looks like I missed some places when doing the Wicket 8 upgrade. I hope I have covered them all in now in https://github.com/inception-project/inception/pull/544

**Comment by reckart** [comment]
@rcffc The title square issue will be fixed along with https://github.com/inception-project/inception/issues/547

I will run a release of our custom RDF4J lucene module so that we can depend here on a stable (non-SNAPSHOT) version. It would be nice if you could meanwhile resolve the conflicts and make sure the build works so that we can wrap this up.

**Comment by reckart** [comment]
I'm having a look at the conflicts...

File: inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/service/ConceptLinkingService.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (reckart) [reply inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/service/ConceptLinkingService.java]: It's a nested tertiary...  although I don't understand why altLabel is preferred over label.

**Comment by reckart** [comment]
It seems to be **mostly** working now (configuration etc.).

However, when I import the wine ontology into a local KB (OWL mapping) with FTS enabled and create a layer with a KB-linked feature that allows classes/instances from any KB/scope - I still only get auto-completions for "Wine Ontology" and "wine", but for nothing else from the KB.

**Comment by rcffc** [comment]
In the Wine ontology, these two are the only items that have used `rdfs:label`. I am having a look at the build.

**Comment by rcffc** [comment]
The changes in KnowledgeBaseServiceImpl cause the MtasDocumentIndexTests fail to load the ApplicationContext.

**Comment by reckart** [comment]
@rcffc yep, I just noticed it too. I'm trying to figure out how to fix it. The problem seems to be that the KB service tries to access the DB before all the entities are available (or that some entities are not available).

**Comment by reckart** [comment]
> In the Wine ontology, these two are the only items that have used rdfs:label. I am having a look at the build.

I see. I wonder if we can maybe hook into/configure the LuceneSail indexing process somehow so that it also allows us to index the IRI suffixes that we show in the UI if no label is available.

**Comment by reckart** [comment]
@rcffc ok, the MTAS tests should work now

**Comment by reckart** [comment]
> I see. I wonder if we can maybe hook into/configure the LuceneSail indexing process somehow so that it also allows us to index the IRI suffixes that we show in the UI if no label is available.

Looks like that is not easily possible. We'd have to make changes directly to the RDF4J Lucene indexing code. The code already stores an "URI" field in the lucene index, but that cannot be queried - at least not via SPARQL as the analyzer used to parse the query tries to remove special characters. We'd get something like this when trying

```
ERROR LuceneIndex - There was a problem evaluating query 'uri:"http://mbugert.de/pets#animal"' for property 'null!
java.lang.IllegalStateException: field "uri" was indexed without position data; cannot run PhraseQuery (phrase=uri:"http mbugert.de pets animal")
```
